pool/web/view/page/page_admin_session.ml

```diff
@@ -817,3 +795,68 @@ let close
        (Page_admin_experiments.Control control)
        experiment
 ;;
+
+let cancel
+  Pool_context.{ language; csrf; _ }
+  experiment
+  (session : Session.t)
+  flash_fetcher
+  =
+  let action =
+    let open Pool_common.Id in
+    Format.asprintf
+      "/admin/experiments/%s/sessions/%s/cancel"
+      (Experiment.Id.value experiment.Experiment.id)
+      (value session.Session.id)
+  in
+  (match session.Session.canceled_at with
+   | Some canceled_at ->
+     Pool_common.(
+       p
+         [ canceled_at
+           |> Utils.Time.formatted_date_time
+           |> Pool_common.Message.sessionalreadycanceled
+           |> Utils.error_to_string language
+           |> txt
+         ])
+   | None ->
+     form
+       ~a:
+         [ a_class [ "stack" ]
+         ; a_method `Post
+         ; a_action (action |> Sihl.Web.externalize_path)
+         ]
+       [ csrf_element csrf ()
+       ; textarea_element
+           ~flash_fetcher
+           language
+           Pool_common.Message.Field.Reason
+       ; span
+           Pool_common.
```

Comment:
may use `let open Pool_common in` at the beginning of the function.

Reply:
Thanks, I've done it everywhere in the file